File: spikes/dist-real-core/node/check_core.mjs
// TS/JS loader — koffi (single prebuilt-friendly FFI dep). Loads the REAL core.
// Binds the actual `citenexus-core` cdylib and calls `citenexus_core_version()`.
import koffi from 'koffi';

const libPath = process.env.CORE_LIB;
const lib = koffi.load(libPath);

// 'str' marshals a NUL-terminated const char*; the returned pointer is a 'static
// crate string (not malloc'd) so there is nothing to free.
const version = lib.func('citenexus_core_version', 'str', []);

const ver = version();
console.log(`[node] citenexus_core_version()=${JSON.stringify(ver)}`);

const expected = (process.env.EXPECT_CORE_VERSION || '').trim();
if (!/^\d+\.\d+\.\d+/.test(ver)) {
  console.error(`[node] MISMATCH: not a semver: ${ver}`);
  process.exit(1);
}
if (expected && ver !== expected) {
  console.error(`[node] MISMATCH: got ${ver}, expected ${expected}`);
  process.exit(1);
}
console.log('[node] OK');

File: spikes/dist-real-core/go/check_core.go
// Go loader — purego (pure Go, NO cgo, no toolchain). Loads the REAL core via
// dlopen/LoadLibrary at runtime and calls `citenexus_core_version()`.
package main

import (
	"fmt"
	"os"
	"regexp"
	"strings"

	"github.com/ebitengine/purego"
)

func main() {
	libPath := os.Getenv("CORE_LIB")
	lib, err := purego.Dlopen(libPath, purego.RTLD_NOW|purego.RTLD_GLOBAL)
	if err != nil {
		fmt.Fprintln(os.Stderr, "[go] dlopen failed:", err)
		os.Exit(1)
	}

	var version func() string
	purego.RegisterLibFunc(&version, lib, "citenexus_core_version")

	ver := version()
	fmt.Printf("[go] citenexus_core_version()=%q\n", ver)

	expected := strings.TrimSpace(os.Getenv("EXPECT_CORE_VERSION"))
	if !regexp.MustCompile(`^\d+\.\d+\.\d+`).MatchString(ver) {
		fmt.Fprintf(os.Stderr, "[go] MISMATCH: not a semver: %q\n", ver)
		os.Exit(1)
	}
	if expected != "" && ver != expected {
		fmt.Fprintf(os.Stderr, "[go] MISMATCH: got %q, expected %q\n", ver, expected)
		os.Exit(1)
	}
	fmt.Println("[go] OK")
}
